src/client/finAuthFetch.ts

```diff
@@ -0,0 +1,144 @@
+import { finCsrfFetch } from './finCsrfFetch'
+
+/**
+ * An enhanced fetch API for a pluggable authentication mechanism
+ *
+ * Transparently handles pre-authentication, authentication fault detection, authentication, and
+ * will replay the requested resource upon successful authentication.
+ *
+ * This fetch function will utilize the finCsrfFetch function internally to execute the request
+ *
+ * Example
+ * ```typescript
+ *
+ * const result = finAuthFetch(request, {
+ * 		authenticator: {
+ * 			isAuthenticated: (response: Response) => response.status !== 401,
+ *			preauthenticate: async (request: RequestInfo) => new Request(request, {headers: {auth_header: '12345'}}),
+ *			authenticate: (response: Response) => return execute_authentication(username, password),
+ *			maxTries: 3
+ * 	}
+ * })
+ * ```
+ *
+ * @link https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch
+ *
+ * @param resource The resource to request.
+ * @param options Optional object containing custom settings and the authenticator.
+ * @returns A promise that resolves to a response object.
+ * @throws Exception on failed authentication
+ */
+export async function finAuthFetch(
+	resource: RequestInfo,
+	options?: RequestInit | RequestInitAuth
+): Promise<Response> {
+	let resp: Response | undefined
+
+	if (options && isAuthInit(options)) {
+		const authenticator = options.authenticator as RequestAuthenticator
+
+		// Pre authenticate if suppprted by authenticator
+		if (authenticator.preauthenticate) {
+			resource = await authenticator.preauthenticate(resource)
+		}
+
+		// Pipe request to the csrf fetch function
+		resp = await finCsrfFetch(resource, options)
+
+		// Check if response is an authentication fault
+		if (!(await authenticator.isAuthenticated?.(resp))) {
+			const maxTries = authenticator.maxTries ?? 3
+			let authCount = 0
+			let authSuccessful = false
+
+			// Authentication generator to generate an authentication call
+			// while we are under the max try threshold
+			const tryAuthentication = async function* () {
+				while (authCount < maxTries) {
+					authCount++
+					yield await authenticator.authenticate(resp as Response)
+				}
+
+				return false
+			}
+
+			// Attempt to authenticate until' generator has reached the max try threshold
+			for await (const didAuthenticate of tryAuthentication()) {
```

Comment:
I see why you're using a generator here but I think a simple for loop without the generator would be easier to follow...

        // Check if response is an authentication fault
	    if (!(await authenticator.isAuthenticated?.(resp))) {
			const maxTries = authenticator.maxTries ?? 3
			let authSuccessful = false

			for (let i = 0; i < maxTries; ++i) {
				const result = await authenticator.authenticate(
					resp as Response
				)

				if (result) {
					authSuccessful = true
					break
				}
			}

			// If we reached the max try threshold and authentication was still unsuccessful
			// throw an exception
			if (!authSuccessful) {
				throw new Error('Authentication Failed')
			}

			// Replay request
			resp = await finCsrfFetch(resource, options)
		}

Reply:
Ahh. Yes, I coded this as if it was using the forEach callback where it doesn't pause during an await. I agree it will be easier to follow that way with the for loop